### ISSUE TITLE
Feature Request: More Comprehensive Twilio/Spoke Phone Number Integration

### ISSUE THREAD

**Issue by mayefsky** [ProgressiveCoders/Spoke#1799]
This is a broad proposal to revamp Twilio phone number management in Spoke, with an eye toward making it easier for non-sophisticated users, and specifically to fix the inability to purchase many numbers at once; it covers a few individual issues that could also possibly be addressed on their own, but also might make sense to think about together as part of a broad project.
**Problem**

- ~Per https://github.com/MoveOnOrg/Spoke/issues/1758, buying multiple numbers at once in Spoke doesn't work currently (at least for me in the standard MoveOn Spoke setup, though I have some poking around still to do). This is a major issue because to my knowledge there is no workaround for this in the Twilio interface either--you have to buy numbers one at a time~ This no longer repros
- If users ever do buy numbers in Twilio (because they don't follow initial instructions 🙃  or because they want a specific number pattern beyond the area code), in the current Spoke world that information seems to get permanently lost, i.e. Spoke is only aware of the numbers it has bought itself (the numbers still work, but you just can't tell what's going on from Spoke as far as how many numbers are in use)
- There's no in-Spoke functionality to assign numbers to the messaging service if you don't do that when you first buy it; or to unassign or delete/stop paying for numbers

Note that some of this is made obsolete/doesn't apply to the on-the-fly messaging service type features that arrived in Spoke 8+. However, as a provider of Spoke to a bunch of smaller campaigns and orgs, there is likely to be a lot of use of one-messaging-service-per-org-but-lots-of-orgs-on-one-build features this cycle.

**Solution**
Ideally on a single page within Spoke Admin:
-List of all phone numbers belonging to the Organization's Twilio Account (whether or not they came from Spoke)
-Ability to buy new numbers, many at a time
--Graceful failure/errors, e.g. if I didn't buy a number, is there some connection issue? Or does Twilio not have numbers from this area code?
-Ability to attach and detach from messaging service, and delete numbers entirely
--Why have a separate detach option? It's somewhat common that users don't buy enough numbers for their first campaigns. If they don't remove those numbers for future campaigns when they've bought more, the old numbers will still get too much usage at first (because Twilio will prioritize continuing to send to existing contacts from existing numbers). But, if the numbers actually had a good response rate, you might want them back in the pool at some point rather than spinning up even more new ones.

Nice to haves:
-Potentially detecting multiple messaging services and allowing numbers to move between them might be a positive thing, for organizations that go over the daily limit for one service but don't want to move to the full on-the-fly style of things
-Allowing people to search more specifically for numbers beyond area code, e.g. area code + exchange

**Context**
Potentially Relevant APIs: https://www.twilio.com/docs/sms/services/api
https://www.twilio.com/docs/sms/services/api/phonenumber-resource
It seems from the documentation like there is no method to mass-buy numbers, but you can buy them one at a time from the API.
Note: I am planning on migrating my build later this week to 9.0 and AWS Lambda, and will see if that resolves the issue with the current mass-buying tool; I am not sure how many others are affected by that issue.

**Comment by maxkc** [comment]
> It seems from the documentation like there is no method to mass-buy numbers, but you can buy them one at a time from the API.

Correct.  API or Console, unless you're looking for >100 numbers at once and they can help.

This is moderately off topic for the Issue here but @ibrand @bobbyoderek do you think it'd be helpful if I put together a page of which area code/numbers/overlays you absolutely not get between now and the election?  VAN talked to them about this a couple weeks ago so I have a bunch of data that I'd be happy to share!

**Comment by ibrand** [comment]
@maxkc Yes I do think that would be helpful. Probably a good thing to email out to Spoke folks with VAN keys or post in Progcode

**Comment by maxkc** [comment]
Great!  Will try to draft something up today or tomorrow.  FWIW I'm still pending on Progcode - not sure if there are any secret strings you can pull on that front though.

**Comment by Frydafly** [comment]
Hey @maxkc checking in- did you manage to get on progcode? I'd love to help with that if not!

**Comment by maxkc** [comment]
> Hey @maxkc checking in- did you manage to get on progcode? I'd love to help with that if not!

I did not, would love a push if possible!  I'll send you an email right now, @Frydafly 😃

**Comment by maxkc** [comment]
@mayefsky should this (now off topic, sorry! my bad) issue be closed?  Is the 'fix' is just to migrate to Spoke v9+?

**Comment by Frydafly** [comment]
I think so, @maxkc 
cc: @mayefsky

**Comment by mayefsky** [comment]
Up to you @Frydafly, I don't think all of the stuff I originally put in this post is resolved for people not using on-the-fly messaging services. From the original post open things:

- If users ever do buy numbers in Twilio (because they don't follow initial instructions 🙃 or because they want a specific number pattern beyond the area code), in the current Spoke world that information seems to get permanently lost, i.e. Spoke is only aware of the numbers it has bought itself (the numbers still work, but you just can't tell what's going on from Spoke as far as how many numbers are in use)
- There's no in-Spoke functionality to assign numbers to the messaging service if you don't do that when you first buy it; or to unassign or delete/stop paying for numbers

It's still the case from my own experience in Spoke 10 (and testing/reconfirming just now) that if you buy a number directly in Twilio, the Spoke phone numbers screen doesn't know it exists and you can't attach it to your service from Spoke; and it's still the case that you can't allocate a number to your messaging service from the Phone Numbers screen if you didn't do so when you first bought it. (It's true that you can now delete numbers from Spoke.)

I don't think the remaining items are that important in the grand scheme of things and in practice I think the current state of the phone number screen is very manageable/sufficient for most folks so I would definitely not be offended if you closed it or deprioritized, but it's not just an 'upgrade' issue unless I'm missing something.

**Comment by Frydafly** [comment]
Thanks for the clarification @mayefsky ! Keeping this open. 
I've had these issues myself, where transitioning into phone number buying in spoke meant that i had to first delete and clear all the numbers I had bought in twilio. Was quite a lot of work. 

On another note, I think that "deleting" numbers in spoke for now only removes them from a messaging service (correct me if I'm wrong, please!) and doesn't actually clear them from twilio, so you're still paying for them. As a newbie, I made the mistake of thinking I was saving money when I wasn't!

**Comment by Frydafly** [comment]
Also, now that we may be moving towards a more twilio-agnostic set up, I wanted to loop in @lperson & @agreenspan24  here who are doing great work and might have some thoughts on this feature request!
